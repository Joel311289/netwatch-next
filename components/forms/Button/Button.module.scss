/* stylelint-disable declaration-block-no-shorthand-property-overrides */
.buttonWrapper {
  @include font($size: 16px, $color: var(--color-dark), $weight: 500);
  @include flex($align: center, $justify: center, $gap: 10px);
  @include box(fit-content, 44px);
  @include no-outline(true);

  font-family: var(--font-text);
  background-color: var(--color-primary);
  border-radius: var(--border-radius-button);
  padding: 0 16px;
  cursor: pointer;
  transition-timing-function: cubic-bezier(0, 0, 1, 1);
  transition-duration: 200ms;
  transition-property: background;
  white-space: nowrap;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,
    box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,
    border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,
    color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;

  &:not([disabled]):hover {
    filter: brightness(87%);
  }

  &[disabled] {
    background-color: var(--color-disabled);
    color: rgb(var(--rgb-color-dark) / 50%) !important;
    cursor: default !important;
    pointer-events: none;

    &.outline {
      opacity: 0.8;
      border-color: var(--color-disabled);
      color: rgb(var(--rgb-color-tooltip) / 60%) !important;

      &:hover {
        background-color: rgb(var(--rgb-color-primary) / 8%) !important;
      }
    }
  }

  &.small {
    padding: 0 10px;
    height: 36px;
    font-size: 0.9rem;
  }

  &.medium {
    border: 3px solid red !important;
  }

  &.large {
    padding: 0 16px;
    height: 50px;
    font-size: 1.2rem;
  }

  &.xlarge {
    padding: 0 20px;
    height: 56px;
    font-size: 1.5rem;
  }

  &.secondary {
    background-color: var(--color-secondary);
    color: rgb(var(--rgb-background-color-primary));

    &:hover {
      background-color: rgb(var(--rgb-background-color-secondary-hover) / 80%);
      filter: brightness(130%);
    }
  }

  &.clear {
    background-color: transparent;
    color: var(--color-primary);

    &:hover {
      background-color: rgb(var(--rgb-background-color-clear-hover) / 8%) !important;
    }
  }

  &.outline {
    background-color: transparent !important;
    color: var(--color-primary);
    border: 1px solid var(--color-primary);

    &:hover {
      background-color: rgb(var(--rgb-background-color-clear-hover) / 8%);
    }
  }
}
