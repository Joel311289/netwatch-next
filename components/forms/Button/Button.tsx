import { MouseEvent, MouseEventHandler } from 'react';
import { ElementChildren, ElementHTML, ElementSkeleton } from '@/types';
import { withSkeleton } from '@/hoc/withSkeleton';
import { withTooltip } from '@/hoc/withTooltip';
import { classes } from '@/utils/helpers';
import styles from '@/components/forms/Button/Button.module.scss';

export type ButtonProps = Partial<typeof defaultProps> &
  ElementHTML &
  ElementChildren &
  ElementSkeleton & {
    size?: 'small' | 'medium' | 'large' | 'xlarge';
    ariaLabel?: string;
    tooltip?: string;
    secondary?: boolean;
    light?: boolean;
    dark?: boolean;
    clear?: boolean;
    outline?: boolean;
    disabled?: boolean;
    rounded?: boolean;
    onClick?: MouseEventHandler<HTMLButtonElement>;
  };

const defaultProps = {
  size: 'medium'
};

const Button = ({
  className,
  children,
  ariaLabel,
  size,
  secondary,
  clear,
  outline,
  disabled,
  tooltip,
  rounded,
  light,
  dark,
  onClick
}: ButtonProps) => {
  const handleClick = (event: MouseEvent<HTMLButtonElement>): void => {
    onClick && onClick(event);
  };

  return (
    <button
      aria-label={ariaLabel || tooltip}
      className={classes(
        styles.wrapper,
        size && styles[size],
        secondary && styles.secondary,
        light && styles.light,
        dark && styles.dark,
        clear && styles.clear,
        outline && styles.outline,
        rounded && styles.rounded,
        className
      )}
      disabled={disabled}
      onClick={handleClick}>
      {children}
    </button>
  );
};

Button.defaultProps = defaultProps;

const ButtonWithTooltip = withTooltip<ButtonProps>(Button);
const ButtonWithSkeleton = withSkeleton<ButtonProps>(ButtonWithTooltip, { variant: 'button' });

export default ButtonWithSkeleton;
