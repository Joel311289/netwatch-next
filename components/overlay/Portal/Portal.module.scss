/** Portal (main component) **/

.wrapper {
  display: block;
}

/** PortalPaper (child component) **/

.backdrop {
  background-color: rgb(var(--rgb-color-backdrop)) !important;
  backdrop-filter: blur(1px);
  transition: all 0.1s cubic-bezier(0.4, 0, 0.2, 1) !important;
  transition-delay: 0.2s !important;
}

.paper {
  box-shadow: none !important;
  background-color: var(--background-color-primary) !important;
}
