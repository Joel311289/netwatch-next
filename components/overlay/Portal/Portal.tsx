import { cloneElement, ComponentPropsWithRef, createContext, useContext } from 'react';
import { PortalProps as PortalPropsMUI } from '@mui/material';
import cn from 'classnames';
import { ElementChildren, ElementHTML, FunctionVoid } from '@/types';
import { withChildrenFiltered } from '@/hoc/withChildrenFiltered';
import { useModal } from '@/hooks/useModal';
import styles from '@/components/overlay/Portal/Portal.module.scss';

/* -------------------------------------------------------------------------- */
/** Context **/
/* -------------------------------------------------------------------------- */

type PortalContextProps = typeof defaultValue & {
  opened?: boolean;
  onChange?: FunctionVoid<boolean>;
};

const defaultValue = {
  opened: false
};

const PortalContext = createContext<PortalContextProps>(defaultValue);

const usePortalContext = () => useContext(PortalContext);

/* -------------------------------------------------------------------------- */
/** PortalHandler (child component) **/
/* -------------------------------------------------------------------------- */

export type PortalHandlerProps = ElementHTML & Required<ElementChildren<JSX.Element>>;

const PortalHandler = ({ children: Child }: PortalHandlerProps) => {
  const { onChange } = usePortalContext();

  return cloneElement(Child, {
    ...Child.props,
    onClick: () => onChange && onChange(true)
  });
};

/* -------------------------------------------------------------------------- */
/** PortalPortal (child component) **/
/* -------------------------------------------------------------------------- */

export type PaperBackdropProps = ComponentPropsWithRef<'div'>;

export type PaperModalProps = {
  container: PortalPropsMUI['container'];
  BackdropProps: Partial<PaperBackdropProps>;
};

type PortalPaperProps = ElementHTML & Required<ElementChildren<JSX.Element>>;

const PortalPaper = ({ children: Child }: PortalPaperProps) => {
  const { opened, onChange } = usePortalContext();
  const modalConfig = {
    container: document.querySelector('.app'),
    BackdropProps: {
      classes: {
        root: styles.backdrop
      }
    }
  };

  return cloneElement(Child, {
    className: cn(styles.portalPaper, Child.props && Child.props.className),
    opened,
    onChange: () => onChange && onChange(false),
    modalConfig,
    ...Child.props
  });
};

/* -------------------------------------------------------------------------- */
/** Portal (main component) **/
/* -------------------------------------------------------------------------- */

export type PortalProps = typeof defaultProps &
  ElementHTML &
  ElementChildren<Array<JSX.Element>> & {
    opened?: boolean;
    onChange?: FunctionVoid<boolean>;
  };

const defaultProps = defaultValue;

const Portal = ({ children, opened, onChange }: PortalProps) => {
  const { isOpened, handleChange } = useModal(opened, onChange);

  return (
    <PortalContext.Provider
      value={{
        opened: isOpened,
        onChange: handleChange
      }}>
      <div className={cn(styles.portalWrapper, 'wrapper')}>{children}</div>
    </PortalContext.Provider>
  );
};

Portal.defaultProps = defaultProps;

const PortalWithChildrenFiltered = withChildrenFiltered(Portal, {
  Handler: PortalHandler,
  Paper: PortalPaper
});

export default PortalWithChildrenFiltered;
