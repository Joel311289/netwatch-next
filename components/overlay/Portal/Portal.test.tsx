import { ReactNode } from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Button from '@/components/forms/Button/Button';
import Portal from '@/components/overlay/Portal/Portal';

const handlerText = 'Open';
const portalContent = 'Content';

const PaperMock = ({ children }: { children: ReactNode }) => (
  <div>
    <span>{children}</span>
  </div>
);

const handler = () => (
  <Portal.Handler>
    <Button>{handlerText}</Button>
  </Portal.Handler>
);
const paper = () => (
  <Portal.Paper>
    <PaperMock>{portalContent}</PaperMock>
  </Portal.Paper>
);

describe('Tests Portal component', () => {
  test('renders component correctly', () => {
    const { container } = render(
      <Portal opened>
        {handler()}
        {paper()}
      </Portal>
    );

    expect(container).toMatchSnapshot();
  });

  test('renders content correctly', () => {
    render(
      <Portal opened>
        {handler()}
        {paper()}
      </Portal>
    );

    expect(screen.getByText(portalContent)).toBeInTheDocument();
  });

  test('calls onChange prop correctly', async () => {
    const handleChange = jest.fn();

    render(
      <Portal onChange={handleChange}>
        {handler()}
        {paper()}
      </Portal>
    );

    userEvent.click(screen.getByText(handlerText));

    await waitFor(() => {
      expect(handleChange).toHaveBeenCalledTimes(1);
    });
  });

  test('opens paper correctly', () => {
    render(
      <Portal>
        {handler()}
        {paper()}
      </Portal>
    );

    userEvent.click(screen.getByText(handlerText));

    expect(screen.getByText(portalContent)).toBeInTheDocument();
  });
});
