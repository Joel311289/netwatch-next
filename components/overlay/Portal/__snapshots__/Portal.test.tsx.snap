// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests Portal component renders component correctly 1`] = `
<div>
  <div
    class="wrapper"
  >
    <button
      class="wrapper medium"
    >
      Open
    </button>
    <div>
      <span>
        Content
      </span>
    </div>
  </div>
</div>
`;
